src/static/js/caretPosition.js

```diff
@@ -0,0 +1,236 @@
+// height, etc. So, we can use the caret line to make calculation necessary to know where is the top
+// of the previous line
+// [2] the line before is part of another rep line. It's possible this line has different margins
+// height. So we have to get the exactly position of the line
+exports.getPositionTopOfPreviousBrowserLine = function(caretLinePosition, rep) {
+  var previousLineTop = caretLinePosition.top - caretLinePosition.height; // [1]
+  var isCaretLineFirstBrowserLine = caretLineIsFirstBrowserLine(caretLinePosition.top, rep);
+
+  // the caret is in the beginning of a rep line, so the previous browser line
+  // is the last line browser line of the a rep line
+  if (isCaretLineFirstBrowserLine) { //[2]
+    var lineBeforeCaretLine = rep.selStart[0] - 1;
+    var firstLineVisibleBeforeCaretLine = getPreviousVisibleLine(lineBeforeCaretLine, rep);
+    var linePosition = getDimensionOfLastBrowserLineOfRepLine(firstLineVisibleBeforeCaretLine, rep);
+    previousLineTop = linePosition.top;
+  }
+  return previousLineTop;
+}
+
+function caretLineIsFirstBrowserLine(caretLineTop, rep)
+{
+  var caretRepLine = rep.selStart[0];
+  var lineNode = rep.lines.atIndex(caretRepLine).lineNode;
+  var firstRootNode = getFirstRootChildNode(lineNode);
+
+  // to get the position of the node we the position of the first char
+  var positionOfFirstRootNode = exports.getPositionOfRepLineAtOffset(firstRootNode, 1);
+  return positionOfFirstRootNode.top === caretLineTop;
+}
+
+// find the first root node, usually it is a text node
+function getFirstRootChildNode(node)
+{
+  if(!node.firstChild){
+    return node;
+  }else{
+    return getFirstRootChildNode(node.firstChild);
+  }
+
+}
+
+function getPreviousVisibleLine(line, rep)
+{
+  if (line < 0) {
+    return 0;
+  }else if (isLineVisible(line, rep)) {
+    return line;
+  }else{
+    return getPreviousVisibleLine(line - 1, rep);
+  }
+}
+
+function getDimensionOfLastBrowserLineOfRepLine(line, rep)
+{
+  var lineNode = rep.lines.atIndex(line).lineNode;
+  var lastRootChildNode = getLastRootChildNode(lineNode);
+
+  // we get the position of the line in the last char of it
+  var lastRootChildNodePosition = exports.getPositionOfRepLineAtOffset(lastRootChildNode.node, lastRootChildNode.length);
+  return lastRootChildNodePosition;
+}
+
+function getLastRootChildNode(node)
+{
+  if(!node.lastChild){
+    return {
+      node: node,
+      length: node.length
+    };
+  }else{
+    return getLastRootChildNode(node.lastChild);
+  }
+}
+
+// here we have two possibilities:
+// [1] The next line is part of the same rep line of the caret line, so we have the same dimensions.
+// So, we can use the caret line to calculate the bottom of the line.
+// [2] the next line is part of another rep line. It's possible this line has different dimensions, so we
+// have to get the exactly dimension of it
+exports.getBottomOfNextBrowserLine = function(caretLinePosition, rep)
```

Comment:
same note of `getPositionTopOfPreviousBrowserLine`

Reply:
actually it's not the same but it's quite similar